# Exploring Hackers News Posts

[Hacker News](https://news.ycombinator.com/) is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

You can find the data set here, but note that it has been reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions. Below are descriptions of the columns:

* id: The unique identifier from Hacker News for the post
* title: The title of the post
* url: The URL that the posts links to, if it the post has a URL
* num_points: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
* num_comments: The number of comments that were made on the post
* author: The username of the person who submitted the post
* created_at: The date and time at which the post was submitted

We're specifically interested in posts whose titles begin with either Ask HN or Show HN. Users submit Ask HN posts to ask the Hacker News community a specific question.

Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll compare these two types of posts to determine the following:

* Do Ask HN or Show HN receive more comments on average?
* Do posts created at a certain time receive more comments on average?

In [1]:
from csv import reader
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)

print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [2]:
headers = hn[:1]
hn = hn[1:]

print(headers)
print("\n")
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']]


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


# Separate all posts into ask_post, show_post and other_posts

In [3]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print("Length of ask posts:", len(ask_posts))
print("Length of show posts:", len(show_posts))
print("Length of other posts:",len(other_posts))

Length of ask posts: 1744
Length of show posts: 1162
Length of other posts: 17194


# Caculate the average number of comments for ask post and show post

In [4]:
total_ask_comments = 0
for row in ask_posts:
    n_comments = int(row[4])
    total_ask_comments += n_comments
    
avg_ask_comments = total_ask_comments/len(ask_posts)

print('The average number of ask post comments is: ', avg_ask_comments)

total_show_comments = 0
for row in show_posts:
    n_comments = int(row[4])
    total_show_comments += n_comments
    
avg_show_comments = total_show_comments/len(show_posts)
print('The average number of show post comments is: ', avg_show_comments)
    

The average number of ask post comments is:  14.038417431192661
The average number of show post comments is:  10.31669535283993


# Ask post receive more comments than show post

In [9]:
import datetime as dt
result_list = []

for row in ask_posts:
    created_at = row[6]
    n_comments = int(row[4])
    result_list.append([created_at,n_comments])    
    
counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    created_at = row[0]
    n_comments = row[1]
    created = dt.datetime.strptime(created_at,"%m/%d/%Y %H:%M")
    hour = created.strftime("%H")
    
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = n_comments
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += n_comments
        

counts_by_hour        
comments_by_hour
    


{'00': 447,
 '01': 683,
 '02': 1381,
 '03': 421,
 '04': 337,
 '05': 464,
 '06': 397,
 '07': 267,
 '08': 492,
 '09': 251,
 '10': 793,
 '11': 641,
 '12': 687,
 '13': 1253,
 '14': 1416,
 '15': 4477,
 '16': 1814,
 '17': 1146,
 '18': 1439,
 '19': 1188,
 '20': 1722,
 '21': 1745,
 '22': 479,
 '23': 543}

In [11]:
avg_by_hour = []

for hour in comments_by_hour:
    avg_by_hour.append([hour,comments_by_hour[hour]/counts_by_hour[hour]])
    
avg_by_hour


[['12', 9.41095890410959],
 ['07', 7.852941176470588],
 ['00', 8.127272727272727],
 ['15', 38.5948275862069],
 ['10', 13.440677966101696],
 ['01', 11.383333333333333],
 ['18', 13.20183486238532],
 ['05', 10.08695652173913],
 ['04', 7.170212765957447],
 ['22', 6.746478873239437],
 ['06', 9.022727272727273],
 ['23', 7.985294117647059],
 ['08', 10.25],
 ['19', 10.8],
 ['03', 7.796296296296297],
 ['02', 23.810344827586206],
 ['21', 16.009174311926607],
 ['09', 5.5777777777777775],
 ['14', 13.233644859813085],
 ['20', 21.525],
 ['11', 11.051724137931034],
 ['17', 11.46],
 ['16', 16.796296296296298],
 ['13', 14.741176470588234]]

In [14]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])

print(swap_avg_by_hour)

[[9.41095890410959, '12'], [7.852941176470588, '07'], [8.127272727272727, '00'], [38.5948275862069, '15'], [13.440677966101696, '10'], [11.383333333333333, '01'], [13.20183486238532, '18'], [10.08695652173913, '05'], [7.170212765957447, '04'], [6.746478873239437, '22'], [9.022727272727273, '06'], [7.985294117647059, '23'], [10.25, '08'], [10.8, '19'], [7.796296296296297, '03'], [23.810344827586206, '02'], [16.009174311926607, '21'], [5.5777777777777775, '09'], [13.233644859813085, '14'], [21.525, '20'], [11.051724137931034, '11'], [11.46, '17'], [16.796296296296298, '16'], [14.741176470588234, '13']]


In [18]:
sorted_swap = sorted(swap_avg_by_hour,reverse=True)

sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [20]:
print("Top 5 Hours for Ask Posts Comments")

for row in sorted_swap[:5]:
    avg  = row[0]
    hr = row[1]
    print("{}: {:.2f} average comments per post".format
         (dt.datetime.strptime(hr,"%H").strftime("%H:%M"), avg)
         )

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

# Conclusion
In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.